tools/syz-declextract/syz-declextract.cpp

```diff
@@ -64,6 +106,230 @@ bool beginsWith(const std::string_view &str, const std::string_view begin) {
+    return field->getType()->isCountAttributedType()
+               ? llvm::dyn_cast<FieldDecl>(
+                     field->getType()->getAs<CountAttributedType>()->getCountExpr()->getReferencedDeclOfCallee())
+                     ->getFieldIndex()
+               : UINT_MAX;
+  }
+
+  bool isFieldVarlen(const QualType &fieldType) {
+    return fieldType->isIncompleteArrayType() ||
+           (fieldType->isConstantArrayType() && llvm::dyn_cast<ConstantArrayType>(fieldType)->getSize().isZero());
+  }
+
+  std::string getStructAttr(const RecordDecl *const recordDecl, ASTContext *context) {
+    if (recordDecl->isStruct()) {
+      for (const auto &item : recordDecl->getAttrs()) {
+        if (item->getKind() == clang::attr::Aligned) {
+          return "align[" + std::to_string(llvm::dyn_cast<AlignedAttr>(item)->getAlignment(*context)) + "]";
+        } else if (item->getKind() == clang::attr::Packed) {
+          return "packed";
+        }
+      }
+    }
+    return "";
+  }
+
+public:
+  RecordExtractor(const SourceManager *const SM) : SM(SM){};
+  std::string getFieldType(const QualType &fieldType, ASTContext *context, const std::string &fieldName,
+                           const std::string &parent = "", bool isSyscallParam = false) {
+    if (contains(fieldName, "ifindex")) {
+      return "ifindex";
+    }
+    const auto &field = fieldType.IgnoreParens();
+    switch (field->getTypeClass()) {
+    case clang::Type::Record:
+      return extractRecord(field->getAsRecordDecl(), context, parent.empty() ? fieldName : parent + "_" + fieldName);
+    case clang::Type::Typedef:
+      return getFieldType(llvm::dyn_cast<TypedefType>(field)->desugar(), context, fieldType.getAsString(), parent,
+                          isSyscallParam);
+    case clang::Type::IncompleteArray: // Defined as type[]
+      return makeArray(getFieldType(llvm::dyn_cast<IncompleteArrayType>(field)->getElementType(), context, fieldName));
+    case clang::Type::ConstantArray: {
+      const auto &array = llvm::dyn_cast<ConstantArrayType>(field);
+      return makeArray(getFieldType(array->getElementType(), context, fieldName), array->getSize().getZExtValue());
+    }
+    case clang::Type::Pointer: {
+      const auto &pointerType = llvm::dyn_cast<PointerType>(field);
+      const auto &pointeeType = pointerType->getPointeeType();
+      const auto &fieldType =
+          pointeeType->isAnyCharacterType() ? "string" : getFieldType(pointeeType, context, fieldName);
+      const auto &ptrDir = pointeeType.isConstQualified() ? "in" : "inout"; // TODO: Infer direction of non-const.
+      return makePtr(ptrDir, fieldType,
+                     parent + "$auto_record" == fieldType); // Checks if the direct parent is the same as the node.
+    }
+    case clang::Type::Builtin: {
+      const auto &type = getSyzType(field.getAsString(), isSyscallParam);
+      return isPadding(fieldName) ? makeConst(isSyscallParam ? "" : type) : type;
```

Comment:
It would be good to merge this into function that refines the type based on field name (as the last case).